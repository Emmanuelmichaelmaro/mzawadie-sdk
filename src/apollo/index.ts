export * from "./ApolloClientManager";
export * from "./client";
export * from "./cache";
export * from "./config";
export * from "./links";
export * from "./utils";
