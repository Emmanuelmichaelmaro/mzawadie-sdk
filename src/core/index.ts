export * from "./auth";
export * from "./consts";
export * from "./createMzawadieClient";
export * from "./states";
export * from "./storage";
export * from "./types";
