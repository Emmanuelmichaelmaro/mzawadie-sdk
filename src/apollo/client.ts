import {
    ApolloCache,
    ApolloClient,
    ApolloLink,
    createHttpLink,
    FetchResult,
    InMemoryCache,
    NormalizedCacheObject,
    Reference,
} from "@apollo/client";
import fetch from "cross-fetch";
import jwtDecode from "jwt-decode";

import { JWTToken } from "../core";
import { storage } from "../core";
import { TypedTypePolicies } from "./apollo-helpers";
import { AuthSDK } from "./auth";
import { auth } from "./auth";
import { RefreshTokenMutation } from "./types";
import { ApolloConfigOptions } from "./utils";

let client: ApolloClient<NormalizedCacheObject>;
let authClient: AuthSDK;
let refreshPromise: ReturnType<AuthSDK["refreshToken"]> | null = null;

type FetchConfig = Partial<{
    /**
     * Enable auto token refreshing. Default to `true`.
     */
    autoTokenRefresh: boolean;
    /**
     * Set a value for skew between local time and token expiration date in
     * seconds (only together with `autoTokenRefresh`). Defaults to `60`.
     */
    tokenRefreshTimeSkew: number;
    /**
     * Refresh token and retry the request when Mzawadie responds with `Unauthorized` error.
     * Defaults to `true`.
     */
    refreshOnUnauthorized: boolean;
}>;

const getTypePolicies = (autologin: boolean): TypedTypePolicies => ({
    Query: {
        fields: {
            authenticated: {
                read(_, { readField, toReference }): boolean {
                    return !!readField(
                        "id",
                        toReference({
                            __typename: "User",
                        })
                    );
                },
            },
            me: {
                read(_, { toReference, canRead }): Reference | undefined | null {
                    const ref = toReference({
                        __typename: "User",
                    });

                    return canRead(ref) ? ref : null;
                },
            },
            authenticating: {
                read(read = autologin && !!storage.getCSRFToken(), { readField }): boolean {
                    if (readField("authenticated")) {
                        return false;
                    }

                    return read;
                },
            },
        },
    },
    User: {
        /**
         * IMPORTANT
         * This works as long as we have 1 User cache object which is the current logged in User.
         * If the client should ever fetch additional Users, this should be removed
         * and the login methods (token create or verify) should be responsible for writing USER query cache manually.
         */
        keyFields: [],
        fields: {
            addresses: {
                merge: false,
            },
        },
    },
});

export const createFetch =
    ({
        autoTokenRefresh = true,
        tokenRefreshTimeSkew = 120,
        refreshOnUnauthorized = true,
    }: FetchConfig = {}) =>
    async (input: RequestInfo, init: RequestInit): Promise<Response> => {
        if (!client) {
            throw new Error(
                "Could not find Mzawadie's client instance. Did you forget to call createMzawadieClient()?"
            );
        }

        let token = storage.getAccessToken();

        if (JSON.parse(init.body?.toString() || "")?.operationName === "refreshToken") {
            return fetch(input, init);
        }

        if (autoTokenRefresh && token) {
            // auto refresh token before provided time skew (in seconds) until it expires
            const expirationTime = (jwtDecode<JWTToken>(token).exp - tokenRefreshTimeSkew) * 1000;

            try {
                if (refreshPromise) {
                    await refreshPromise;
                } else if (Date.now() >= expirationTime) {
                    // refreshToken automatically updates token in storage
                    refreshPromise = authClient.refreshToken();

                    await refreshPromise;
                }
            } catch (e) {
            } finally {
                refreshPromise = null;
            }
            token = storage.getAccessToken();
        }

        if (token) {
            init.headers = {
                ...init.headers,
                "authorization-bearer": token,
            };
        }

        if (refreshOnUnauthorized && token) {
            const response = await fetch(input, init);
            const data: FetchResult = await response.clone().json();
            const isUnauthenticated = data?.errors?.some(
                (error) => error.extensions?.exception.code === "ExpiredSignatureError"
            );
            let refreshTokenResponse: FetchResult<
                RefreshTokenMutation,
                Record<string, any>,
                Record<string, any>
            >;

            if (isUnauthenticated) {
                try {
                    if (refreshPromise) {
                        refreshTokenResponse = await refreshPromise;
                    } else {
                        refreshPromise = authClient.refreshToken();

                        refreshTokenResponse = await refreshPromise;
                    }

                    if (refreshTokenResponse.data?.tokenRefresh?.token) {
                        // check if mutation returns a valid token after refresh and retry the request
                        return createFetch({
                            autoTokenRefresh: false,
                            refreshOnUnauthorized: false,
                        })(input, init);
                    } else {
                        // after Mzawadie returns ExpiredSignatureError status and token refresh fails
                        // we log out the user and return the failed response
                        await authClient.logout();
                    }
                } catch (e) {
                } finally {
                    refreshPromise = null;
                }
            }

            return response;
        }

        return fetch(input, init);
    };

export const createApolloClient1 = (
    apiUrl: string,
    autologin: boolean
): ApolloClient<NormalizedCacheObject> => {
    const httpLink = createHttpLink({
        fetch: createFetch(),
        uri: apiUrl,
        credentials: "include",
    });

    const cache = new InMemoryCache({
        typePolicies: getTypePolicies(autologin),
    });

    client = new ApolloClient({
        cache,
        link: httpLink,
    });

    /**
     * Refreshing token code should stay under apollo/auth.ts To get this method available,
     * we need to call "auth()" here. refreshToken mutation doesn't require channel, so it
     * doesn't have to be populated with value.
     */
    authClient = auth({ apolloClient: client, channel: "" });

    return client;
};

/**
 * Creates Apollo client.
 * @param cache Cache used by created Apollo client.
 * @param links Links used by created Apollo client.
 * @param options Rest options, which might be passed to Apollo client.
 */
export function createApolloClient2(
    cache: ApolloCache<any>,
    links: ApolloLink[],
    options?: ApolloConfigOptions
) {
    return new ApolloClient({
        ...options,
        cache,
        link: ApolloLink.from(links),
    });
}
