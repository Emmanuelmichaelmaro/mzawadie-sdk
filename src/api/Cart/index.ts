import { Config } from "../../apollo";
import { ApolloClientManager } from "../../apollo/ApolloClientManager";
import { MzawadieState, MzawadieStateLoaded } from "../../core/state";
import { IMzawadieStateSummeryPrices, StateItems } from "../../core/state/types";
import { LocalStorageManager } from "../../data";
import { ErrorListener } from "../../helpers";
import { ICheckoutModel } from "../../helpers/LocalStorageHandler";
import { JobsManager } from "../../jobs";
import { ErrorCartTypes } from "../../jobs/Cart";
import { IDiscount, IItems, IShippingPrice, ISubtotalPrice, ITotalPrice } from "./types";
import { sortCheckoutLines } from "./utils";

export class MzawadieCartAPI extends ErrorListener {
    loaded: boolean;

    items: IItems;

    totalPrice: ITotalPrice;

    subtotalPrice: ISubtotalPrice;

    shippingPrice: IShippingPrice;

    discount?: IDiscount;

    private apolloClientManager: ApolloClientManager;

    private jobsManager: JobsManager;

    private localStorageManager: LocalStorageManager;

    private mzawadieState: MzawadieState;

    // temporary solution, might change in future
    private config: Config;

    constructor(
        localStorageManager: LocalStorageManager,
        apolloClientManager: ApolloClientManager,
        mzawadieState: MzawadieState,
        jobsManager: JobsManager,
        config: Config
    ) {
        super();
        this.mzawadieState = mzawadieState;
        this.localStorageManager = localStorageManager;
        this.apolloClientManager = apolloClientManager;
        this.jobsManager = jobsManager;
        this.config = config;

        this.loaded = false;

        this.jobsManager.attachErrorListener("cart", this.fireError);

        this.mzawadieState.subscribeToChange(StateItems.CHECKOUT, (checkout: ICheckoutModel) => {
            this.items = checkout?.lines
                ?.filter((line: any) => line.quantity > 0)
                .sort(sortCheckoutLines);
        });
        this.mzawadieState.subscribeToChange(
            StateItems.SUMMARY_PRICES,
            (summaryPrices: IMzawadieStateSummeryPrices) => {
                const { totalPrice, subtotalPrice, shippingPrice, discount } = summaryPrices || {};
                this.totalPrice = totalPrice;
                this.subtotalPrice = subtotalPrice;
                this.shippingPrice = shippingPrice;
                this.discount = discount;
            }
        );
        this.mzawadieState.subscribeToChange(StateItems.LOADED, (loaded: MzawadieStateLoaded) => {
            this.loaded = loaded.checkout && loaded.summaryPrices;
        });
    }

    addItem = async (variantId: string, quantity: number) => {
        // 1. save in local storage
        this.localStorageManager.addItemToCart(variantId, quantity);

        // 2. save online if possible (if checkout id available)
        if (this.mzawadieState.checkout?.lines) {
            const { data, error } = await this.apolloClientManager.getRefreshedCheckoutLines(
                this.mzawadieState.checkout.lines,
                this.config.channel
            );

            if (error) {
                this.fireError(error, ErrorCartTypes.SET_CART_ITEM);
            } else {
                this.localStorageManager.getHandler().setCheckout({
                    ...this.mzawadieState.checkout,
                    lines: data,
                });
            }
        }
        if (this.mzawadieState.checkout?.id) {
            this.jobsManager.addToQueue("cart", "setCartItem");
            return {
                pending: true,
            };
        }
        return {
            pending: false,
        };
    };

    removeItem = async (variantId: string) => {
        // 1. save in local storage
        this.localStorageManager.removeItemFromCart(variantId);
        // 2. save online if possible (if checkout id available)
        if (this.mzawadieState.checkout?.lines) {
            const { data, error } = await this.apolloClientManager.getRefreshedCheckoutLines(
                this.mzawadieState.checkout.lines,
                this.config.channel
            );

            if (error) {
                this.fireError(error, ErrorCartTypes.SET_CART_ITEM);
            } else {
                this.localStorageManager.getHandler().setCheckout({
                    ...this.mzawadieState.checkout,
                    lines: data,
                });
            }
        }
        if (this.mzawadieState.checkout?.id) {
            this.jobsManager.addToQueue("cart", "setCartItem");
            return {
                pending: true,
            };
        }
        return {
            pending: false,
        };
    };

    subtractItem = async (variantId: string) => {
        // 1. save in local storage
        this.localStorageManager.subtractItemFromCart(variantId);

        // 2. save online if possible (if checkout id available)
        if (this.mzawadieState.checkout?.lines) {
            const { data, error } = await this.apolloClientManager.getRefreshedCheckoutLines(
                this.mzawadieState.checkout.lines,
                this.config.channel
            );

            if (error) {
                this.fireError(error, ErrorCartTypes.SET_CART_ITEM);
            } else {
                this.localStorageManager.getHandler().setCheckout({
                    ...this.mzawadieState.checkout,
                    lines: data,
                });
            }
        }
        if (this.mzawadieState.checkout?.id) {
            this.jobsManager.addToQueue("cart", "setCartItem");
            return {
                pending: true,
            };
        }
        return {
            pending: false,
        };
    };

    updateItem = async (variantId: string, quantity: number) => {
        // 1. save in local storage
        this.localStorageManager.updateItemInCart(variantId, quantity);

        // 2. save online if possible (if checkout id available)
        if (this.mzawadieState.checkout?.lines) {
            const { data, error } = await this.apolloClientManager.getRefreshedCheckoutLines(
                this.mzawadieState.checkout.lines,
                this.config.channel
            );

            if (error) {
                this.fireError(error, ErrorCartTypes.SET_CART_ITEM);
            } else {
                this.localStorageManager.getHandler().setCheckout({
                    ...this.mzawadieState.checkout,
                    lines: data,
                });
            }
        }
        if (this.mzawadieState.checkout?.id) {
            this.jobsManager.addToQueue("cart", "setCartItem");
            return {
                pending: true,
            };
        }
        return {
            pending: false,
        };
    };
}
