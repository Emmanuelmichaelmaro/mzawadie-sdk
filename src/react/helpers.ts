import React from "react";

import { MzawadieAPI } from "../api";
import { ConfigInput } from "../apollo";
import { MzawadieContext } from "./context";

export function useMzawadieClient(): MzawadieAPI {
    const mzawadie = React.useContext(MzawadieContext);

    if (!mzawadie) {
        throw new Error(
            "Could not find mzawadie's apollo client in the context. " +
                "Did you forget to wrap the root component in a <MzawadieProvider>?"
        );
    }

    return mzawadie.api;
}

export function useMzawadieConfig(): ConfigInput {
    const mzawadie = React.useContext(MzawadieContext);

    if (!mzawadie) {
        throw new Error(
            "Could not find mzawadie's apollo client in the context. " +
                "Did you forget to wrap the root component in a <MzawadieProvider>?"
        );
    }

    return mzawadie.config;
}
