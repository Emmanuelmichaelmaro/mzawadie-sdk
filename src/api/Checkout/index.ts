// @ts-nocheck
import { Config } from "../../apollo";
import { PaymentGateway } from "../../apollo/fragments/gqlTypes/PaymentGateway";
import { MzawadieState, MzawadieStateLoaded } from "../../core/state";
import { StateItems } from "../../core/state/types";
import { ErrorListener } from "../../helpers";
import { ICheckoutModel, IPaymentModel } from "../../helpers/LocalStorageHandler";
import { JobsManager } from "../../jobs";
import { PromiseRunResponse } from "../types";
import {
    DataErrorCheckoutTypes,
    FunctionErrorCheckoutTypes,
    IAddress,
    IAvailableShippingMethods,
    ICheckout,
    IPayment,
    IPromoCodeDiscount,
    CreatePaymentInput,
    CompleteCheckoutInput,
} from "./types";

type CheckoutResponse = PromiseRunResponse<DataErrorCheckoutTypes, FunctionErrorCheckoutTypes>;

export class MzawadieCheckoutAPI extends ErrorListener {
    loaded: boolean;

    checkout?: ICheckout;

    promoCodeDiscount?: IPromoCodeDiscount;

    billingAsShipping?: boolean;

    selectedShippingAddressId?: string;

    selectedBillingAddressId?: string;

    availableShippingMethods?: IAvailableShippingMethods;

    availablePaymentGateways?: PaymentGateway[];

    payment?: IPayment;

    private mzawadieState: MzawadieState;

    private jobsManager: JobsManager;

    // temporary solution, might change in future
    private config: Config;

    constructor(mzawadieState: MzawadieState, jobsManager: JobsManager, config: Config) {
        super();
        this.mzawadieState = mzawadieState;
        this.jobsManager = jobsManager;
        this.config = config;

        this.loaded = false;

        this.mzawadieState.subscribeToChange(StateItems.CHECKOUT, (checkout: ICheckoutModel) => {
            const {
                id,
                token,
                email,
                shippingAddress,
                billingAddress,
                selectedShippingAddressId,
                selectedBillingAddressId,
                billingAsShipping,
                availablePaymentGateways,
                availableShippingMethods,
                shippingMethod,
                promoCodeDiscount,
            } = checkout || {};
            this.checkout = {
                billingAddress,
                email,
                id,
                shippingAddress,
                shippingMethod,
                token,
            };
            this.selectedShippingAddressId = selectedShippingAddressId;
            this.selectedBillingAddressId = selectedBillingAddressId;
            this.availablePaymentGateways = availablePaymentGateways;
            this.availableShippingMethods = availableShippingMethods;
            this.billingAsShipping = billingAsShipping;
            this.promoCodeDiscount = {
                discountName: promoCodeDiscount?.discountName,
                voucherCode: promoCodeDiscount?.voucherCode,
            };
        });
        this.mzawadieState.subscribeToChange(StateItems.PAYMENT, (payment: IPaymentModel) => {
            const { id, token, gateway, creditCard, total } = payment || {};
            this.payment = {
                creditCard,
                gateway,
                id,
                token,
                total,
            };
        });
        this.mzawadieState.subscribeToChange(StateItems.LOADED, (loaded: MzawadieStateLoaded) => {
            this.loaded = loaded.checkout && loaded.payment;
        });
    }

    setShippingAddress = async (shippingAddress: IAddress, email: string): CheckoutResponse => {
        const checkoutId = this.mzawadieState.checkout?.id;
        const alteredLines = this.mzawadieState.checkout?.lines?.map((item) => ({
            quantity: item!.quantity,
            variantId: item?.variant!.id,
        }));

        if (alteredLines && checkoutId) {
            const { data, dataError } = await this.jobsManager.run("checkout", "setShippingAddress", {
                channel: this.config.channel,
                checkoutId,
                email,
                selectedShippingAddressId: shippingAddress.id,
                shippingAddress,
            });

            return {
                data,
                dataError,
                pending: false,
            };
        }
        if (alteredLines) {
            const { data, dataError } = await this.jobsManager.run("checkout", "createCheckout", {
                channel: this.config.channel,
                email,
                lines: alteredLines,
                selectedShippingAddressId: shippingAddress.id,
                shippingAddress,
            });

            return {
                data,
                dataError,
                pending: false,
            };
        }
        return {
            functionError: {
                error: new Error("You need to add items to cart before setting shipping address."),
                type: FunctionErrorCheckoutTypes.ITEMS_NOT_ADDED_TO_CART,
            },
            pending: false,
        };
    };

    setBillingAddress = async (billingAddress: IAddress, email?: string): CheckoutResponse => {
        const checkoutId = this.mzawadieState.checkout?.id;
        const isShippingRequiredForProducts = this.mzawadieState.checkout?.lines
            ?.filter((line) => line.quantity > 0)
            .some(({ variant }) => variant.product?.productType.isShippingRequired);
        const alteredLines = this.mzawadieState.checkout?.lines?.map((item) => ({
            quantity: item!.quantity,
            variantId: item?.variant!.id,
        }));

        if (isShippingRequiredForProducts && checkoutId && this.checkout?.shippingAddress) {
            const { data, dataError } = await this.jobsManager.run("checkout", "setBillingAddress", {
                billingAddress,
                billingAsShipping: false,
                checkoutId,
                selectedBillingAddressId: billingAddress.id,
            });

            return {
                data,
                dataError,
                pending: false,
            };
        }
        if (isShippingRequiredForProducts) {
            return {
                functionError: {
                    error: new Error(
                        "You need to set shipping address before setting billing address."
                    ),
                    type: FunctionErrorCheckoutTypes.SHIPPING_ADDRESS_NOT_SET,
                },
                pending: false,
            };
        }
        if (!isShippingRequiredForProducts && email && checkoutId && alteredLines) {
            const { data, dataError } = await this.jobsManager.run(
                "checkout",
                "setBillingAddressWithEmail",
                {
                    billingAddress,
                    checkoutId,
                    email,
                    selectedBillingAddressId: billingAddress.id,
                }
            );

            return {
                data,
                dataError,
                pending: false,
            };
        }
        if (!isShippingRequiredForProducts && email && alteredLines) {
            const { data, dataError } = await this.jobsManager.run("checkout", "createCheckout", {
                billingAddress,
                channel: this.config.channel,
                email,
                lines: alteredLines,
                selectedBillingAddressId: billingAddress.id,
            });

            return {
                data,
                dataError,
                pending: false,
            };
        }
        if (!isShippingRequiredForProducts && !email) {
            return {
                functionError: {
                    error: new Error(
                        "You need to provide email when products do not require shipping before setting billing address."
                    ),
                    type: FunctionErrorCheckoutTypes.EMAIL_NOT_SET,
                },
                pending: false,
            };
        }
        return {
            functionError: {
                error: new Error("You need to add items to cart before setting billing address."),
                type: FunctionErrorCheckoutTypes.ITEMS_NOT_ADDED_TO_CART,
            },
            pending: false,
        };
    };

    setBillingAsShippingAddress = async (): PromiseRunResponse<
        DataErrorCheckoutTypes,
        FunctionErrorCheckoutTypes
    > => {
        const checkoutId = this.mzawadieState.checkout?.id;

        if (checkoutId && this.checkout?.shippingAddress) {
            const { data, dataError } = await this.jobsManager.run("checkout", "setBillingAddress", {
                billingAddress: this.checkout.shippingAddress,
                billingAsShipping: true,
                checkoutId,
                selectedBillingAddressId: this.checkout?.shippingAddress.id,
            });

            return {
                data,
                dataError,
                pending: false,
            };
        }
        return {
            functionError: {
                error: new Error("You need to set shipping address before setting billing address."),
                type: FunctionErrorCheckoutTypes.SHIPPING_ADDRESS_NOT_SET,
            },
            pending: false,
        };
    };

    setShippingMethod = async (shippingMethodId: string): CheckoutResponse => {
        const checkoutId = this.mzawadieState.checkout?.id;

        if (checkoutId) {
            const { data, dataError } = await this.jobsManager.run("checkout", "setShippingMethod", {
                checkoutId,
                shippingMethodId,
            });
            return {
                data,
                dataError,
                pending: false,
            };
        }
        return {
            functionError: {
                error: new Error("You need to set shipping address before setting shipping method."),
                type: FunctionErrorCheckoutTypes.SHIPPING_ADDRESS_NOT_SET,
            },
            pending: false,
        };
    };

    addPromoCode = async (promoCode: string): CheckoutResponse => {
        const checkoutId = this.mzawadieState.checkout?.id;

        if (checkoutId) {
            const { data, dataError } = await this.jobsManager.run("checkout", "addPromoCode", {
                checkoutId,
                promoCode,
            });

            return {
                data,
                dataError,
                pending: false,
            };
        }
        return {
            functionError: {
                error: new Error("You need to set shipping address before modifying promo code."),
                type: FunctionErrorCheckoutTypes.SHIPPING_ADDRESS_NOT_SET,
            },
            pending: false,
        };
    };

    removePromoCode = async (promoCode: string): CheckoutResponse => {
        const checkoutId = this.mzawadieState.checkout?.id;

        if (checkoutId) {
            const { data, dataError } = await this.jobsManager.run("checkout", "removePromoCode", {
                checkoutId,
                promoCode,
            });

            return {
                data,
                dataError,
                pending: false,
            };
        }
        return {
            functionError: {
                error: new Error("You need to set shipping address before modifying promo code."),
                type: FunctionErrorCheckoutTypes.SHIPPING_ADDRESS_NOT_SET,
            },
            pending: false,
        };
    };

    createPayment = async (input: CreatePaymentInput): CheckoutResponse => {
        const checkoutId = this.mzawadieState.checkout?.id;
        const amount = this.mzawadieState.summaryPrices?.totalPrice?.gross.amount;

        if (checkoutId && amount !== null && amount !== undefined) {
            const { data, dataError } = await this.jobsManager.run("checkout", "createPayment", {
                ...input,
                amount,
                checkoutId,
            });
            return {
                data,
                dataError,
                pending: false,
            };
        }
        return {
            functionError: {
                error: new Error("You need to set billing address before creating payment."),
                type: FunctionErrorCheckoutTypes.SHIPPING_ADDRESS_NOT_SET,
            },
            pending: false,
        };
    };

    completeCheckout = async (input?: CompleteCheckoutInput): CheckoutResponse => {
        const checkoutId = this.mzawadieState.checkout?.id;

        if (checkoutId) {
            const { data, dataError } = await this.jobsManager.run("checkout", "completeCheckout", {
                ...input,
                checkoutId,
            });
            return {
                data,
                dataError,
                pending: false,
            };
        }
        return {
            functionError: {
                error: new Error("You need to set shipping address before creating payment."),
                type: FunctionErrorCheckoutTypes.SHIPPING_ADDRESS_NOT_SET,
            },
            pending: false,
        };
    };
}
