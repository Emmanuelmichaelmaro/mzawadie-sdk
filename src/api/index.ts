import { ApolloClient } from "@apollo/client";

import { Config } from "../apollo";
import { ApolloClientManager } from "../apollo/ApolloClientManager";
import { MzawadieState } from "../core/state";
import { LocalStorageManager } from "../data";
import { LocalStorageHandler } from "../helpers/LocalStorageHandler";
import { JobsManager } from "../jobs";
import APIProxy from "./APIProxy";
import { AuthAPI } from "./Auth";
import { MzawadieCartAPI } from "./Cart";
import { MzawadieCheckoutAPI } from "./Checkout";
import { CategoriesAPI } from "./categories/categories";
import { CollectionsAPI } from "./collections";
import { ProductsAPI } from "./products";

export * from "./Checkout";
export * from "./Cart";

export class MzawadieAPI {
    auth: AuthAPI;

    checkout: MzawadieCheckoutAPI;

    cart: MzawadieCartAPI;

    categories: CategoriesAPI;

    collections: CollectionsAPI;

    products: ProductsAPI;

    /**
     * @deprecated Please do not use it anymore. Reference to API Proxy will be removed in future.
     * Now it just exists for legacy React hooks, which also will be removed.
     */
    legacyAPIProxy: APIProxy;

    constructor(
        client: ApolloClient<any>,
        apiProxy: APIProxy,
        config: Config,
        onStateUpdate?: () => any
    ) {
        this.legacyAPIProxy = apiProxy;

        const localStorageHandler = new LocalStorageHandler();

        const apolloClientManager = new ApolloClientManager(client);

        const jobsManager = new JobsManager(localStorageHandler, apolloClientManager);

        const mzawadieState = new MzawadieState(
            config,
            localStorageHandler,
            apolloClientManager,
            jobsManager
        );

        const localStorageManager = new LocalStorageManager(localStorageHandler, mzawadieState);

        if (onStateUpdate) {
            mzawadieState.subscribeToNotifiedChanges(onStateUpdate);
        }

        this.auth = new AuthAPI(mzawadieState, jobsManager, config);

        this.checkout = new MzawadieCheckoutAPI(mzawadieState, jobsManager, config);

        this.cart = new MzawadieCartAPI(
            localStorageManager,
            apolloClientManager,
            mzawadieState,
            jobsManager,
            config
        );

        this.categories = new CategoriesAPI(client);

        this.collections = new CollectionsAPI(client, config);

        this.products = new ProductsAPI(client, config);

        this.legacyAPIProxy.attachAuthListener((authenticated: any) => {
            if (!authenticated) {
                localStorageHandler.setCheckout({});
                localStorageHandler.setPayment({});
                localStorageHandler.setJobs(null);
            }
        });
    }
}
