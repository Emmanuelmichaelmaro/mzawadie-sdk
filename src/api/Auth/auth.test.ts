// @ts-nocheck
import { InMemoryCache } from "@apollo/client";

import { AuthAPI } from ".";
import { setupRecording, setupAPI } from "../../../testUtils/api";
import { defaultConfig } from "../../apollo";
import { ApolloClientManager } from "../../apollo/ApolloClientManager";
import { API_CHANNEL } from "../../core";
import { MzawadieState } from "../../core/state";
import { LocalStorageHandler } from "../../helpers";
import { JobsManager } from "../../jobs";
import APIProxy from "../APIProxy";

setupRecording();

describe("Auth API", () => {
    let authAPI: AuthAPI;
    let cache: InMemoryCache;
    let apiProxy: APIProxy;

    // Auth tests have custom recording matcher setup in the ./setup.ts.
    // Thanks to that, changing user email and password will not trigger
    // test to fail.
    // const context = setupRecording();
    // const mzawadie = setupMzawadieClient();

    // beforeEach(() => {
    //     const { server } = context.polly;
    //     setupPollyMiddleware(server);
    // });

    beforeAll(async (done) => {
        const { client, cache: apiCache, apiUrl } = await setupAPI();

        const channel = API_CHANNEL;

        cache = apiCache;

        const config = {
            ...defaultConfig,
            apiUrl,
            channel,
        };

        const localStorageHandler = new LocalStorageHandler();

        const apolloClientManager = new ApolloClientManager(client);

        const jobsManager = new JobsManager(localStorageHandler, apolloClientManager);

        const mzawadieState = new MzawadieState(
            config,
            localStorageHandler,
            apolloClientManager,
            jobsManager
        );

        authAPI = new AuthAPI(mzawadieState, jobsManager, config);

        apiProxy = new APIProxy(client);

        done();
    });

    it("Returns error if credentials are invalid", async () => {
        const signInResult = await authAPI.signIn("sdk@example.com", "test1");

        expect(!!signInResult.data?.id).toBe(false);
        expect(signInResult.dataError?.error?.length).toBe(1);
        expect(signInResult.functionError).toBe(undefined);
        expect(signInResult.pending).toBe(false);

        expect(!!authAPI.user?.id).toBe(false);
        expect(authAPI.authenticated).toBe(false);
        expect(authAPI.loaded).toBe(true);
        expect(!!authAPI.token).toBe(false);
    });

    it("Can sign in", async () => {
        const signInResult = await authAPI.signIn("sdk@example.com", "test");

        expect(signInResult.data).toMatchSnapshot();

        expect(!!signInResult.data?.user?.id).toBe(true);
        expect(signInResult.dataError).toBe(undefined);
        expect(signInResult.functionError).toBe(undefined);
        expect(signInResult.pending).toBe(false);

        expect(!!authAPI.user?.id).toBe(true);
        expect(authAPI.authenticated).toBe(true);
        expect(authAPI.loaded).toBe(true);
        expect(!!authAPI.token).toBe(true);
    });

    it("Does not cache mutations", async (done) => {
        await authAPI
            .signIn("sdk@example.com", "test")
            .then(() =>
                apiProxy.setPassword({
                    email: "sdk@example.com",
                    password: "test2345678",
                    token: "5hr-73a06b70fd6ad8ab3913",
                })
            )
            .catch(() =>
                apiProxy.setPasswordChange({
                    newPassword: "test2345678",
                    oldPassword: "test2345678",
                })
            )
            .catch(() => {
                const extractedCache = cache.extract();
                expect(extractedCache.ROOT_MUTATION).toBeUndefined();

                done();
            });
    });

    it("Can sign out", async () => {
        expect(!!authAPI.user?.id).toBe(true);
        expect(authAPI.authenticated).toBe(true);
        expect(authAPI.loaded).toBe(true);
        expect(!!authAPI.token).toBe(true);

        const signOutResult = await authAPI.signOut();

        expect(signOutResult.pending).toBe(false);

        expect(!!authAPI.user?.id).toBe(false);
        expect(authAPI.authenticated).toBe(false);
        expect(authAPI.loaded).toBe(true);
        expect(!!authAPI.token).toBe(false);
    });
});
