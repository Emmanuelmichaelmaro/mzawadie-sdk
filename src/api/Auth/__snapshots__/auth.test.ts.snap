// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth API Can sign in 1`] = `undefined`;

exports[`Auth API Returns error if credentials are invalid 1`] = `undefined`;
